crates/validation/src/lib.rs

```diff
@@ -186,3 +200,95 @@ pub async fn validate<D: Drivers>(
         inferences,
     }
 }
+
+/// Parses the image inspection json into an `assemble::PortMap`, which can be used to create the shard template.
+fn parse_image_inspection(
+    image: &str,
+    image_inspections: &[tables::ImageInspection],
+) -> Result<PortMap, Error> {
+    let row = image_inspections
+        .iter()
+        .find(|r| r.image == image)
+        .ok_or_else(|| Error::ImageInspectFailed {
+            image: image.to_owned(),
+            error: anyhow::anyhow!("image inspection results missing"),
+        })?;
+    if let Some(err) = row.inspect_error.as_ref() {
+        return Err(Error::ImageInspectFailed {
+            image: image.to_owned(),
+            error: anyhow::format_err!(err.clone()),
+        });
+    }
+
+    let deserialized: Vec<InspectJson> =
+        serde_json::from_slice(&row.inspect_output).map_err(|err| {
+            let output_str = String::from_utf8_lossy(&row.inspect_output);
+            eprintln!("deserializing docker inspect output failed: {}", output_str);
+            Error::ImageInspectFailed {
+                image: image.to_owned(),
+                error: anyhow::Error::from(err),
+            }
+        })?;
+
+    if deserialized.len() != 1 {
+        return Err(Error::ImageInspectFailed {
+            image: image.to_owned(),
+            error: anyhow::anyhow!("expected 1 image, got {}", deserialized.len()),
+        });
+    }
+    let mut ports = BTreeMap::new();
+    for (port_config, _) in deserialized[0].config.exposed_ports.iter() {
+        // We're unable to support UDP at this time.
+        if port_config.ends_with("/udp") {
+            continue;
```

Comment:
Would we want to return an error here instead of filtering these out?

Reply:
I don't think so. I'm considering cases where you might use `FROM some-external-image` where `some-external-image` happens to expose a UDP port for some optional features, and you can't really control it. And in the event that we want to add support for UDP, I don't see a problem with going from "UDP ports are silently ignored" to "exposed UDP ports are private by default"